extmod/uasyncio: Add global exception handling

kevinkk525 opened the pull request:
Support global exception handling to uasyncio according to CPython error handling: https://docs.python.org/3/library/asyncio-eventloop.html#error-handling-api

Adds the Loop methods (also changing all loop methods to be staticmethos or classmethods since there is only one loop) and a default exception handler. 

This is especially interesting since this new version of uasyncio doesn't throw exceptions to the caller of "loop.run_forever()" and therefore exceptions in other tasks are only printed to repl, where the user might never see it since the device will be deployed without logging the repl output. 
With a global exception handling a user can catch those exceptions and send them by mqtt or log them to a file on the device. 

The implementation preallocates a context dictionary so in case of an exception there shouldn't be any RAM allocations.

The used approach is compatible to CPython except for 2 problems:
1) There is no way to log the Exception traceback because "sys.print_exception" only prints the traceback to the repl. So there is no way to actually log the traceback, which would be very helpful. Hopefully this can be implemented. I understand that this might cause RAM allocation but a user might decide to use it anyway in a custom exception handler because it makes debugging a lot easier if you know in what file and line the error occured.
Edit: Apparently there is a way so that's fine.
2) In CPython the exception handler is called once the task is finished which created the task that threw an uncaught exception, whereas in UPy the exception handler is called immediately when the exception is thrown. This makes a difference in the following testcase but is generally just a minor difference that shouldn't cause any abnormal behaviour.

```
async def test_catch_uncaught_exception():
    # can't work with a local return value because the exception handler runs after
    # the current coroutine is finished in CPython. Works in UPy though.
    res = False

    async def fail():
        raise TypeError("uncaught exception")

    def handle_exception(loop, context):
        # context["message"] will always be there; but context["exception"] may not
        print(context)
        print(context["message"])
        print(context["exception"])
        msg = context.get("exception", context["message"])
        if mp:
            print("Caught: {}{}".format(type(context["exception"]), msg))
        else:
            print("Caught: {}".format({msg}))
        nonlocal res
        print("res is", res)
        res = True
        print("done")

    t = asyncio.create_task(fail())
    loop = asyncio.get_event_loop()
    loop.set_exception_handler(handle_exception)
    await asyncio.sleep(0.1)
    await asyncio.sleep(0.1)
    print("coro done")
    return res
```

kevinkk525 committed:
Add global exception handling

Support global exception handling to uasyncio according to CPython error handling: https://docs.python.org/3/library/asyncio-eventloop.html#error-handling-api

Adds the Loop methods (also changing all loop methods to be staticmethos or classmethods since there is only one loop) and a default exception handler. 

This is especially interesting since this new version of uasyncio doesn't throw exceptions to the caller of "loop.run_forever()" and therefore exceptions in other tasks are only printed to repl, where the user might never see it since the device will be deployed without logging the repl output. 
With a global exception handling a user can catch those exceptions and send them by mqtt or log them to a file on the device. 

The implementation preallocates a context dictionary so in case of an exception there shouldn't be any RAM allocations.

The used approach is compatible to CPython except for 2 problems:
1) There is no way to log the Exception traceback because "sys.print_exception" only prints the traceback to the repl. So there is no way to actually log the traceback, which would be very helpful. Hopefully this can be implemented. I understand that this might cause RAM allocation but a user might decide to use it anyway in a custom exception handler because it makes debugging a lot easier if you know in what file and line the error occured.
2) In CPython the exception handler is called once the task is finished which created the task that threw an uncaught exception, whereas in UPy the exception handler is called immediately when the exception is thrown. This makes a difference in the following testcase but is generally just a minor difference that shouldn't cause any abnormal behaviour.

```
async def test_catch_uncaught_exception():
    # can't work with a local return value because the exception handler runs after
    # the current coroutine is finished in CPython. Works in UPy though.
    res = False

    async def fail():
        raise TypeError("uncaught exception")

    def handle_exception(loop, context):
        # context["message"] will always be there; but context["exception"] may not
        print(context)
        print(context["message"])
        print(context["exception"])
        msg = context.get("exception", context["message"])
        if mp:
            print("Caught: {}{}".format(type(context["exception"]), msg))
        else:
            print("Caught: {}".format({msg}))
        nonlocal res
        print("res is", res)
        res = True
        print("done")

    t = asyncio.create_task(fail())
    loop = asyncio.get_event_loop()
    loop.set_exception_handler(handle_exception)
    await asyncio.sleep(0.1)
    await asyncio.sleep(0.1)
    print("coro done")
    return res
```

tve commented on extmod/uasyncio/core.py:
```diff
@@ -215,20 +234,41 @@ def run(coro):
+
+    @classmethod
+    def call_exception_handler(cls, context):
+        exc_handler(cls, context)
```
As far as I can tell from the CPython docs this should be `_exc_handler`

kevinkk525 replied:
Where did you read that? 
"exc_handler" isn't exposed as part of the API. In fact, I should have named it differently to reflect that it is only an internal variable not to be used or changed externally.

tve replied:
Hey, I have no experience with this exception handling, so it's very possible that I'm misunderstanding. What I saw on the page you linked to is:

>loop.default_exception_handler(context)¶
>Default exception handler.
>This is called when an exception occurs and no exception handler is set. This can be called by a custom exception handler that wants to defer to the default handler behavior.

My understanding of this last sentence is that if I register a custom exception handler (thus `exc_handler == my_custom_handler`) and in this custom handler I want to just call the standard default handler I can call `loop.default_exception_handler()`, thus the latter should call `_exc_handler`. (More elegant would be to rename `_exc_handler` to `default_exception_handler`.) But maybe I'm not understanding it quite right (I find the python docs more ambiguous than I'm used to for this type of stuff...)

tve replied:
Whooops! Now I see my mistake, my comment should have been attached to line 265 (4 lines up).

kevinkk525 replied:
Ah, indeed. That is a mistake on my part. it should be _exec_handler here.

tve reviewed (commented):
Note a workaround to your issue #1:
```
                    errbuf = io.BytesIO(1400)
                    sys.print_exception(e, errbuf)
                    errbuf = errbuf.getvalue()
```
Not saying this is great or that it doesn't have issues (it does!).

kevinkk525 commented:
Thanks @tve for the workaround with the errbuf. It works (and that's nice) but it is probably not a good way to allocate a fixed length buffer.
However, your solution made me think.. What do you think about this:

```
errbuf=io.StringIO()
sys.print_exception(e, errbuf)
errbuf=errbuf.getvalue()
```

errbuf should only allocate what is needed as far as I know.

kevinkk525 committed:
Update core.py

kevinkk525 committed:
rename _exc_handler variables

to have a more clear distinction between the default exc_handler and the _exc_handler variable.
Also make exc_handler an internal variable.

dpgeorge commented on extmod/uasyncio/core.py:
```diff
@@ -215,20 +234,41 @@ def run(coro):
 
 class Loop:
-    def create_task(self, coro):
+    @staticmethod
```
Note that making the functions all staticmethod's will increase RAM usage because each use of the staticmethod decorator allocates a block in the heap.  Since RAM is precious I think it's worth trying to optimise it here and not use `staticmethod` if at all possible.

kevinkk525 replied:
Hmm I see, I changed everything back to not use staticmethods.
The downside is, that every exception now first needs to create a Loop() instance before the exception handler can be called.

dpgeorge replied:
In #5841 I've removed the need to create a `Loop()` instance.

dpgeorge commented on extmod/uasyncio/core.py:
```diff
@@ -215,20 +234,41 @@ def run(coro):
+
+    @classmethod
+    def default_exception_handler(cls, context):
+        _default_exc_handler(cls, context)
```
How about putting the implementation of `_default_exc_handler` right here?  That would avoid defining this function above and save RAM (each function definition uses RAM).

kevinkk525 committed:
Update core.py

kevinkk525 committed:
Update core.py

dpgeorge commented on extmod/uasyncio/core.py:
```diff
@@ -228,6 +239,24 @@ def run_until_complete(self, aw):
+        _exc_handler = handler
+
+    def get_exception_handler(self):
+        return _exc_handler if _exc_handler is not None else self.default_exception_handler
```
By initialising `_ext_handler` like this:
```python
_exc_handler = Loop.default_exception_handler
```
this function can simply return `_exc_handler`.

dpgeorge commented on extmod/uasyncio/core.py:
```diff
@@ -228,6 +239,24 @@ def run_until_complete(self, aw):
+        if _exc_handler is None:
+            self.default_exception_handler(context)
+        else:
+            _exc_handler(self, context)
```
This can be simplified to just the last line if `_ext_handler` has `default_exception_handler` set as the default/init value.

kevinkk525 committed:
extmod/uasyncio/core: Exception handling without Loop

kevinkk525 commented:
I moved everything into the Loop and removed all "self" of the Exception handling.

dpgeorge commented:
See #5855 for an extension of this PR with docs and tests.
